File: ios/AgentChat/Services/AttachmentUploader.swift
import Foundation

struct UploadedAttachment: Codable {
    let id: String
    let filename: String
    let mimeType: String
    let byteCount: Int
}

@MainActor
final class AttachmentUploader {
    static let shared = AttachmentUploader()
    private init() {}

    /// Upload an attachment's data to the MCP server. Returns the server-side metadata.
    func upload(attachment: Attachment, host: String, mcpPort: Int, authToken: String) async throws -> UploadedAttachment {
        guard let data = AttachmentStore.shared.loadData(for: attachment) else {
            throw UploadError.fileNotFound
        }

        guard let url = URL(string: "http://\(host):\(mcpPort)/api/attachments") else {
            throw UploadError.invalidURL
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.httpBody = data
        request.setValue(attachment.mimeType, forHTTPHeaderField: "Content-Type")
        request.setValue(attachment.filename, forHTTPHeaderField: "X-Filename")
        request.setValue("Bearer \(authToken)", forHTTPHeaderField: "Authorization")

        let (responseData, response) = try await URLSession.shared.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 else {
            let statusCode = (response as? HTTPURLResponse)?.statusCode ?? 0
            throw UploadError.serverError("Upload failed with status \(statusCode)")
        }

        return try JSONDecoder().decode(UploadedAttachment.self, from: responseData)
    }
}

enum UploadError: LocalizedError {
    case fileNotFound
    case invalidURL
    case serverError(String)

    var errorDescription: String? {
        switch self {
        case .fileNotFound: return "Attachment file not found"
        case .invalidURL: return "Invalid MCP server URL"
        case .serverError(let msg): return msg
        }
    }
}

File: mobile/src/services/AttachmentUploader.ts
import * as FileSystem from 'expo-file-system';
import { Attachment, UploadedAttachment } from '../types/attachment';

/**
 * Uploads a local attachment to the MCP server's /api/attachments endpoint.
 * Uses expo-file-system uploadAsync for reliable binary uploads on React Native.
 */
export async function uploadAttachment(
  attachment: Attachment,
  host: string,
  mcpPort: number,
  authToken: string,
): Promise<UploadedAttachment> {
  if (!attachment.localUri) {
    throw new Error(`No local file for attachment: ${attachment.filename}`);
  }

  const url = `http://${host}:${mcpPort}/api/attachments`;

  const result = await FileSystem.uploadAsync(url, attachment.localUri, {
    httpMethod: 'POST',
    uploadType: FileSystem.FileSystemUploadType.BINARY_CONTENT,
    headers: {
      'Content-Type': attachment.mimeType,
      'X-Filename': attachment.filename,
      'Authorization': `Bearer ${authToken}`,
    },
  });

  if (result.status < 200 || result.status >= 300) {
    throw new Error(`Upload failed (${result.status}): ${result.body || 'Unknown error'}`);
  }

  return JSON.parse(result.body) as UploadedAttachment;

  if (!response.ok) {
    const body = await response.text().catch(() => '');
    throw new Error(`Upload failed (${response.status}): ${body || response.statusText}`);
  }

  return response.json() as Promise<UploadedAttachment>;
}
